### PR TITLE
5324 create new role moderator

### PR DESCRIPTION
**Pull request by zaziemo**
The people who are assigned the moderator role should be able to see, mark and delete reported comments/posts. They also should get an email when a new post/comment is reported. Mostly everything works fine, only the email's language doesn't change according to the person's language settings (it takes the language settings of the first one).

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -150,7 +157,7 @@ def gon_set_current_user
     return unless user_signed_in?
     a_ids = session[:a_ids] || []
     user = UserPresenter.new(current_user, a_ids)
-    gon.push({:user => user})
+    gon.push({user: user})
```
Comment (diaspora-code-review) [inline app/controllers/application_controller.rb]: Redundant curly braces around a hash parameter.

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -88,7 +88,14 @@ def set_locale
...
   def redirect_unless_admin
     unless current_user.admin?
-      redirect_to stream_url, :notice => 'you need to be an admin to do that'
+      redirect_to stream_url, notice: "you need to be an admin to do that"
+      return
+    end
+  end
+
+  def redirect_unless_admin_or_moderator
+    unless current_user.moderator? || current_user.admin?
```
Comment (diaspora-code-review) [inline app/controllers/application_controller.rb]: Use a guard clause instead of wrapping the code inside a conditional expression.

File: app/controllers/report_controller.rb [hunk 3]
```diff
@@ -4,7 +4,7 @@
 
 class ReportController < ApplicationController
   before_filter :authenticate_user!
-  before_filter :redirect_unless_admin, :except => [:create]
+  before_filter :redirect_unless_admin_or_moderator, except: [:create]
```
Comment (diaspora-code-review) [inline app/controllers/report_controller.rb]: Prefer `before_action` over `before_filter`.

File: app/mailers/report_mailer.rb [hunk 4]
```diff
@@ -1,30 +1,32 @@
...
-    Role.admins.each do |role|
-      person = Person.find(role.person_id)
-      if person.local?
-        user = User.find_by_id(person.owner_id)
-        unless user.user_preferences.exists?(:email_type => :someone_reported)
-          resource[:email] = user.email
-          format(resource)
-        end
+    person = Person.find(role.person_id)
+    if person.local?
```
Comment (diaspora-code-review) [inline app/mailers/report_mailer.rb]: Use a guard clause instead of wrapping the code inside a conditional expression.

File: app/models/user.rb [hunk 5]
```diff
@@ -489,6 +489,10 @@ def admin?
     Role.is_admin?(self.person)
   end
 
+  def moderator?
+    Role.moderator?(self.person)
```
Comment (diaspora-code-review) [inline app/models/user.rb]: Redundant `self` detected.

**Comment by Flaburgan** [comment]
Awesome <3

File: app/assets/templates/header_tpl.jst.hbs [hunk 6]
```diff
@@ -94,6 +94,9 @@
                 {{#if current_user.admin}}
                   <li><a href="/admins/dashboard">{{t "header.admin"}}</a></li>
                 {{/if}}
+                {{#if current_user.moderator}}
```
Comment (troessner) [inline app/assets/templates/header_tpl.jst.hbs]: Let's rather use obvious predicate methods here:

```Ruby
current_user.moderator?
```

File: app/assets/templates/header_tpl.jst.hbs [hunk 7]
```diff
@@ -94,6 +94,9 @@
...
+                  <li><a href="/report">Moderator</a></li>
```
Comment (troessner) [inline app/assets/templates/header_tpl.jst.hbs]: Shouldn't this be internationalized to support multiple languages?

File: app/controllers/application_controller.rb [hunk 8]
```diff
@@ -35,7 +35,7 @@ def default_serializer_options
   end
 
   def ensure_http_referer_is_set
-    request.env['HTTP_REFERER'] ||= '/'
+    request.env["HTTP_REFERER"] ||= "/"
```
Comment (troessner) [inline app/controllers/application_controller.rb]: Was it intentional to clean this up in this PR as well? Personally I prefer "one PR == one feature" and then have cleanup PR's afterwards. Otherwise it's hard to see what change is coming from where and why.

File: app/controllers/application_controller.rb [hunk 9]
```diff
@@ -68,11 +68,11 @@ def ensure_page
...
 
   def set_diaspora_header
-    headers['X-Diaspora-Version'] = AppConfig.version_string
+    headers["X-Diaspora-Version"] = AppConfig.version_string
 
     if AppConfig.git_available?
-      headers['X-Git-Update'] = AppConfig.git_update if AppConfig.git_update.present?
-      headers['X-Git-Revision'] = AppConfig.git_revision if AppConfig.git_revision.present?
+      headers["X-Git-Update"] = AppConfig.git_update if AppConfig.git_update.present?
+      headers["X-Git-Revision"] = AppConfig.git_revision if AppConfig.git_revision.present?
```
Comment (troessner) [inline app/controllers/application_controller.rb]: That's what I meant above - by doing those cleanup tasks as well it's harder for me, the reviewer, to actually .....review this PR.:)

File: app/mailers/report_mailer.rb [hunk 10]
```diff
@@ -1,30 +1,32 @@
 class ReportMailer < ActionMailer::Base
-  default :from => AppConfig.mail.sender_address
+  default from: AppConfig.mail.sender_address
 
-  def new_report(type, id)
+  def self.new_report(type, id)
```
Comment (troessner) [inline app/mailers/report_mailer.rb]: Personally I'd prefer the use of "module_function" instead of the "self" syntax but maybe that's just me.:)

File: app/models/role.rb [hunk 11]
```diff
@@ -17,6 +18,14 @@ def self.add_admin(person)
     find_or_create_by(person_id: person.id, name: "admin")
   end
 
+  def self.moderator?(person)
+    exists?(person_id: person.id, name: "moderator")
+  end
```
Comment (jhass) [inline app/models/role.rb]: Maybe we should just consider all admins moderators? So the scope would become `where(name: %w(moderator admin))`, this would become `moderators.exists?(person_id: person.id)`, the filter would become `redirect_unless_moderator` and reports would be mailed to `Role.moderators.map`

File: app/mailers/report_mailer.rb [hunk 12]
```diff
@@ -1,30 +1,32 @@
...
+    [Role.admins + Role.moderators].flatten.map {|role| super(type, id, role) }
```
Comment (troessner) [inline app/mailers/report_mailer.rb]: You can use Enumerable#flat_map here which is more concise. Bonus points: it's significantly faster as well -> https://github.com/JuanitoFatas/fast-ruby#enumerablemaparrayflatten-vs-enumerableflat_map-code

File: app/mailers/report_mailer.rb [hunk 13]
```diff
@@ -1,30 +1,32 @@
...
+  end
+
+  def new_report(type, id, role)
```
Comment (troessner) [inline app/mailers/report_mailer.rb]: How about using keyword arguments here to improve the readability of the method signature?

File: app/mailers/report_mailer.rb (hunk 10 above)
Comment (jhass) [inline app/mailers/report_mailer.rb]: Not applicable here, `#new_report` and `::new_report` are intentionally different methods.

File: app/mailers/report_mailer.rb [hunk 14]
```diff
@@ -1,30 +1,32 @@
...
-      if person.local?
-        user = User.find_by_id(person.owner_id)
-        unless user.user_preferences.exists?(:email_type => :someone_reported)
-          resource[:email] = user.email
-          format(resource)
-        end
+    person = Person.find(role.person_id)
+    if person.local?
+      user = User.find_by_id(person.owner_id)
+      unless user.user_preferences.exists?(email_type: :someone_reported)
```
Comment (troessner) [inline app/mailers/report_mailer.rb]: Slight LoD violation: http://devblog.avdi.org/2011/07/05/demeter-its-not-just-a-good-idea-its-the-law/

File: app/mailers/report_mailer.rb (hunk 13 above)
Comment (jhass) [inline app/mailers/report_mailer.rb]: Note it's called through ActionMailer via the `super` call above.

File: app/models/user.rb (hunk 5 above)
Comment (troessner) [inline app/models/user.rb]: :+1:  ;)

File: app/presenters/user_presenter.rb [hunk 15]
```diff
@@ -6,16 +6,16 @@ def initialize(user, aspects_ids)
...
-    self.user.person.as_api_response(:backbone).update(
-      { :notifications_count => notifications_count,
-        :unread_messages_count => unread_messages_count,
-        :admin => admin,
-        :aspects => aspects,
-        :services => services,
-        :following_count => self.user.contacts.receiving.count,
-        :configured_services => self.configured_services,
-      }
+  def to_json(options={})
```
Comment (troessner) [inline app/presenters/user_presenter.rb]: Instead of using hashes I'd use keyword arguments here as well.

File: app/mailers/report_mailer.rb (hunk 14 above)
Comment (jhass) [inline app/mailers/report_mailer.rb]: But it wasn't introduced by this PR, so I would welcome fixing it in a separate PR.

File: app/presenters/user_presenter.rb (hunk 15 above)
Comment (jhass) [inline app/presenters/user_presenter.rb]: The API is mandated by the json stdlib.

File: app/presenters/user_presenter.rb [hunk 16]
```diff
@@ -24,14 +24,14 @@ def services
...
+    user.services.map(&:provider)
   end
 
   def aspects
     @aspects ||= begin
                    aspects = AspectPresenter.as_collection(user.aspects)
-                   no_aspects = self.aspects_ids.empty?
-                   aspects.each{ |a| a[:selected] = no_aspects || self.aspects_ids.include?(a[:id].to_s) }
+                   no_aspects = aspects_ids.empty?
+                   aspects.each {|a| a[:selected] = no_aspects || aspects_ids.include?(a[:id].to_s) }
```
Comment (troessner) [inline app/presenters/user_presenter.rb]: This "each" is pretty odd. Or maybe it's because I don't know the domain language. What is an "aspect" when it comes to an user?

File: app/presenters/user_presenter.rb [hunk 17]
```diff
@@ -46,4 +46,8 @@ def unread_messages_count
   def admin
     user.admin?
   end
+
+  def moderator
```
Comment (troessner) [inline app/presenters/user_presenter.rb]: I'd make that a properly syntaxed predicate method (with a "?"). Additionally, you could use the Forwardable module in ruby to use implicit delegation instead of explicit one even though I realize this might be a little overkill here.

File: app/presenters/user_presenter.rb (hunk 16 above)
Comment (jhass) [inline app/presenters/user_presenter.rb]: Finding a way to refactor this is completely out of scope of this PR

File: app/views/report/index.html.haml [hunk 18]
```diff
@@ -1,7 +1,8 @@
 .container
   .row
     .col-md-3
-      = render partial: "admins/admin_bar"
+      - if @current_user.admin?
+        = render partial: "admins/admin_bar"
```
Comment (troessner) [inline app/views/report/index.html.haml]: My rails days are long gone but i think you can just do

```Ruby
= render "admins/admin_bar"
````

File: spec/controllers/report_controller_spec.rb [hunk 19]
```diff
@@ -2,128 +2,180 @@
...
+      it "succeeds" do
+        put :update, id: @message.id, type: "post"
+        expect(response.status).to eq(302)
+        expect(Report.where(reviewed: true, item_id: @message.id, item_type: "post")).to be_truthy
+      end
+    end
+    context "mark comment report as moderator" do
+      before do
+        Role.add_moderator(alice.person)
+      end
```
Comment (troessner) [inline spec/controllers/report_controller_spec.rb]: Newline missing.

File: spec/controllers/report_controller_spec.rb [hunk 20]
```diff
@@ -2,128 +2,180 @@
...
-        expect(Report.where(:reviewed => true, :item_id => @comment.id, :item_type => 'comment')).to be_truthy
+        expect(Report.where(reviewed: true, item_id: @comment.id, item_type: "comment")).to be_truthy
+      end
+    end
+
+    context "mark post report as moderator" do
+      before do
+        Role.add_moderator(alice.person)
+      end
+      it "succeeds" do
```
Comment (troessner) [inline spec/controllers/report_controller_spec.rb]: Newline missing.

File: app/presenters/user_presenter.rb (hunk 17 above)
Comment (jhass) [inline app/presenters/user_presenter.rb]: We're inside Rails, no need for forwardable, http://api.rubyonrails.org/classes/Module.html#method-i-delegate

File: spec/controllers/report_controller_spec.rb [hunk 21]
```diff
@@ -2,128 +2,180 @@
...
-        expect(Report.where(:reviewed => false, :item_id => @message.id, :item_type => 'post')).to be_truthy
+        expect(Report.where(reviewed: false, item_id: @message.id, item_type: "post")).to be_truthy
       end
     end
-    context 'destroy comment as user' do
-      it 'is behind redirect_unless_admin' do
-        delete :destroy, :id => @comment.id, :type => 'comment'
+    context "destroy comment as user" do
+      it "is behind redirect_unless_admin_or_moderator" do
+        delete :destroy, id: @comment.id, type: "comment"
```
Comment (troessner) [inline spec/controllers/report_controller_spec.rb]: I'd rather use lazy loaded "lets" instead of boring old instance variables here and everywhere else: http://betterspecs.org/#let

File: app/mailers/report_mailer.rb [hunk 22]
```diff
@@ -1,30 +1,32 @@
...
+        resource[:email] = user.email
+        format(resource)
```
Comment (jhass) [inline app/mailers/report_mailer.rb]: Does wrapping this into `I18n.with_locale(user.language) do` help regarding the localization issue?

File: spec/models/role_spec.rb [hunk 23]
```diff
@@ -45,6 +45,13 @@
     end
   end
 
+  describe ".add_moderator" do
```
Comment (troessner) [inline spec/models/role_spec.rb]: I'd rather have a generic "add" method that takes :moderator as an argument. Otherwise you'll pretty quick bloat your models with repetitive methods like that.

File: app/presenters/user_presenter.rb (hunk 17 above)
Comment (troessner) [inline app/presenters/user_presenter.rb]: Ah, yeah. Like I said, my rails is rusty.:)

File: app/mailers/report_mailer.rb (hunk 14 above)
Comment (troessner) [inline app/mailers/report_mailer.rb]: :+1:

File: app/assets/templates/header_tpl.jst.hbs (hunk 7 above)
Comment (zaziemo) [inline app/assets/templates/header_tpl.jst.hbs]: yeah, absolutely. We kept it for later and then forgot :)

File: app/controllers/application_controller.rb (hunk 8 above)
Comment (zaziemo) [inline app/controllers/application_controller.rb]: Yes, this was intentional because we thought to clean up the obvious things rubocop suggested. We didn't think about it being confusing.

File: app/controllers/application_controller.rb (hunk 2 above)
Comment (zaziemo) [inline app/controllers/application_controller.rb]: @troessner or @jhass could someone of you explain how that guard clause should look like? We didn't manage to change the code accordingly.

Comment (jhass) [inline app/controllers/application_controller.rb]:
```rb
def redirect_unless_admin_or_moderator
  return if current_user.moderator? || current_user.admin?
```

File: app/models/role.rb (hunk 11 above)
Comment (zaziemo) [inline app/models/role.rb]: @jhass we also reflected some alternatives to the current design... we'll think it over again. I find it a little bit confusing to only mention moderators but include admins. But @Realtin also was pro including both roles somehow.

File: app/mailers/report_mailer.rb (hunk 22 above)
Comment (zaziemo) [inline app/mailers/report_mailer.rb]: Yeah! That works. Thx, @jhass!

Comment (jhass) [inline app/mailers/report_mailer.rb]: Great! Would love a spec that ensures it stays that way ;)

File: app/controllers/application_controller.rb (hunk 2 above)
Comment (troessner) [inline app/controllers/application_controller.rb]: But that would drop the message below @jhass 

@zaziemo Think about what happens when you would like to add another role, say "super_admin"? Then the method name would need to change to "redirect_unless_admin_or_moderator_or_super_admin". And then you'd like to another role and all hell would break loose :)

I'd rather go with something flexible like this:

```Ruby
def authorize
  return if user_allowed?
  redirect_to stream_url, notice: "you don't have the necessary rights to do that"
end

def user_allowed?
  current_user.moderator? || current_user.admin?
end
```

Overall this is still not the best design since i'd rather go with something like "user has_many roles" or a gem that abstracts away role management but I guess thats out of the scope of this PR..;)

File: app/mailers/report_mailer.rb [hunk 24]
```diff
@@ -1,18 +1,21 @@
...
+      id:   id
     }
-
-    Role.admins.each do |role|
-      person = Person.find(role.person_id)
-      if person.local?
-        user = User.find_by_id(person.owner_id)
-        unless user.user_preferences.exists?(:email_type => :someone_reported)
+    person = Person.find(role.person_id)
+    if person.local?
```
Comment (diaspora-code-review) [inline app/mailers/report_mailer.rb]: Use a guard clause instead of wrapping the code inside a conditional expression.

File: app/assets/templates/header_tpl.jst.hbs (hunk 6 above)
Comment (jhass) [inline app/assets/templates/header_tpl.jst.hbs]: That's a Handlebars template rendered on the client side, not Ruby ;)

**Comment by Realtin** [comment]
Ok this should be ready to merge. 
I squashed the refactor commits. do you want more squashed? (I'm getting a hang of it now..)

File: app/models/role.rb [hunk 25]
```diff
@@ -17,6 +18,14 @@ def self.add_admin(person)
     find_or_create_by(person_id: person.id, name: "admin")
   end
 
+  def self.moderator?(person)
+    exists?(person_id: person.id, name: "moderator")
```
Comment (jhass) [inline app/models/role.rb]: This should be `moderators.exists?(person_id: person.id)` and then the `|| current_user.admin?` above shouldn't be needed anymore.

Comment (jhass) [inline app/models/role.rb]: And then three specs would be awesome, `expect(Role.moderator?(admin_person)).to be_true`, `expect(Role.moderator?(moderator_person)).to be_true`, `expect(Role.moderator?(normal_person)).to be_false`

**Comment by Realtin** [comment]
We added the specs and refactored the method. what do you think?

**Comment by jhass** [comment]
I think we're almost done here. I'd like the three specs for `Role.moderator?(some_person)` specifically, but the ones for the scope are good too, so don't drop them.

**Comment by zaziemo** [comment]
@jhass do you mean in the same document but outside the scope?

File: spec/models/role_spec.rb [hunk 26]
```diff
@@ -22,6 +23,20 @@
...
+
+      it "should include moderators" do
+        expect(Role.moderators).to include(moderator_role)
+      end
+
+      it "should not include normal users" do
+        expect(Role.moderators).to_not include(person)
+      end
+    end
   end
```
Comment (jhass) [inline spec/models/role_spec.rb]: No. I'm saying that these tests are for the scope `moderators` and it's good to have them, but I'd also like to see them for `Role.moderator?(person)`, like we have specs for `Role.is_admin?(person)` below.

**Comment by zaziemo** [comment]
@jhass should be ready now. We rebased to the current develop state. 
Travis fails are caused by the cucumber tests.

**Comment by jhass** [comment]
Merged as d38741d5ebc8ac8a4e0696f172713626ed1f3a6f c2c6ed5dea50a8ac50141c404afda9e17f7a3126 15b186518c47665f3765bf175aa9f94068f05e8c 098c30c2b7dbfd6138de591b3ca379cb4ef9d87e 0b420d0308046bfac3f63f341f52a9e2806afa90 b2dc77e1e6cbef4702f9ba23605d775126eea306 bc75371b7a95f8ead5f463caf4ef7fe7db32c021 3a3c8817358c3487886ce272c58ac359fa51cb37

Thank you!

**Comment by Flaburgan** [comment]
Nice! Thank you!

**Comment by Flaburgan** [comment]
@zaziemo can you also edit [the wiki](https://wiki.diasporafoundation.org/FAQ_for_pod_maintainers#What_are_roles_and_how_do_I_use_them.3F_.2F_Make_yourself_an_admin) to indicate to podmin how to turn a user into a moderator?

**Comment by zaziemo** [comment]
@Flaburgan sure. How do I indicate that the feature will only be available with the next release?

**Comment by jhass** [comment]
Usually we do wiki updates only with the releases.
